src/lib.rs

```diff
@@ -0,0 +1,548 @@
+    if max_offset > encoded_length - 2 { return err_invalid_offset }
+
+    // Get non-zero offsets and determine the number of segments
+    let mut nr_segments: u8 = 0;  // `nr_segments` is in [0, 15]
+    let mut offsets: Vec<u32> = Vec::with_capacity(15);
+    for val in all_offsets.iter().filter(|&n| *n != 0) {
+        offsets.push(*val);
+        nr_segments += 1u8;
+    }
+
+    // First offset must always be 64
+    if offsets[0] != 64 { return err_invalid_offset }
+
+    // Ensure we have a final ending offset at the end of the data
+    offsets.push(u32::try_from(encoded_length).unwrap());
+
+    // Ensure offsets are in increasing order
+    let mut last: u32 = 0;
+    for val in offsets.iter() {
+        if *val <= last {
+            return err_invalid_offset
+        }
+        last = *val;
+    }
+
+    // Check the samples per pixel is conformant
+    let samples_per_px: u8 = nr_segments / bytes_per_pixel;
+    match samples_per_px {
+        1 => {},
+        3 => {},
+        _ => return err_invalid_nr_samples
+    }
+
+    // Watch for overflow here; u32 * u32 -> u64
+    let expected_length = usize::try_from(
+        px_per_sample * u32::from(bytes_per_pixel * samples_per_px)
+    ).unwrap();
+
+    // Pre-allocate a vector for the decoded frame
+    let mut frame = vec![0u8; expected_length];
+
+    /*
+    Example
+    -------
+    RLE encoded data is ordered like this (for 16-bit, 3 sample):
+      Segment: 1     | 2     | 3     | 4     | 5     | 6
+               R MSB | R LSB | G MSB | G LSB | B MSB | B LSB
+
+    A segment contains only the MSB or LSB parts of all the sample pixels
+
+    To minimise the amount of array manipulation later, and to make things
+    faster we interleave each segment in a manner consistent with a planar
+    configuration of 1 (and maintain big endian byte ordering):
+      All red samples             | All green samples           | All blue
+      Pxl 1   Pxl 2   ... Pxl N   | Pxl 1   Pxl 2   ... Pxl N   | ...
+      MSB LSB MSB LSB ... MSB LSB | MSB LSB MSB LSB ... MSB LSB | ...
+    */
+
+    // Decode each segment and place it into the vector
+    // ------------------------------------------------
+    let pps = usize::try_from(px_per_sample).unwrap();
+    // Concatenate sample planes into a frame
+    for sample in 0..samples_per_px {  // 0 or (0, 1, 2)
+        // Sample offset
+        let so = usize::from(sample * bytes_per_pixel) * pps;
+
+        // Interleave the segments into a sample plane
+        for byte_offset in 0..bytes_per_pixel {  // 0, [1, 2, 3, ..., 7]
+            // idx should be in range [0, 23], but max is 15
+            let idx = usize::from(sample * bytes_per_pixel + byte_offset);
+
+            // offsets[idx] is u32 -> usize not guaranteed
+            let start = usize::try_from(offsets[idx]).unwrap();
+            let end = usize::try_from(offsets[idx + 1]).unwrap();
+
+            // Decode the segment into the frame
+            _decode_segment_plane(
```

Comment:
_decode_segment_plane

Note that we're just sort of assuming that the segment decoded into the correct length, see if that can be checked